pkg/validations/base.go

```diff
@@ -37,6 +38,24 @@ func RunValidations(request reconcile.Request, obj client.Object, kind string, i
 		return
 	}
 
+	// If controller has no replicas clear existing metrics and
+	// do not run any validations
+
+	objValue := reflect.Indirect(reflect.ValueOf(obj))
+	spec := objValue.FieldByName("Spec")
+	if spec.IsValid() {
+		replicas := spec.FieldByName("Replicas")
+		if replicas.IsValid() {
+			numReplicas, ok := replicas.Interface().(*int32)
+
+			// clear labels if we fail to get a value for numReplicas, or if value is <= 0
+			if !ok || numReplicas == nil || *numReplicas <= 0 {
+				engine.DeleteMetrics(promLabels)
+				return
+			}
+		}
+	}
```

Comment:
@rrati

So is the idea here to just assume that objects with a `replicas` field are controllers?

Or do we want to explicitly check the object's kind to be a deployment, replicaset, statefulset, deploymentconfig, etc?

I think the former is probably fine, but hypothetically there may be k8s kinds with a `replicas` field that aren't controllers.

Reply:
A concern:

With the code as-is, I think someone could add a bogus `replicas` field  with value `0` to a `Service`, or a `ServiceAccount` (for example), and then bypass any checks.

Reply:
Hm, I just tested this, and k8s will just ignore invalid properties (e.g. `replicas`) in the stored YAML definition of applied objects. So this actually may not be a concern.

Reply:
I don't know of any API objects that have a replicas field which are not controllers.  I think it's safe to assume that is the case unless we find an edge case otherwise.  It's possible that maybe a custom resource of some kind could have a replicas field that isn't a controller, but until we hit such a thing I don't think we should plan for it.